Development/nmos/jwk_utils.cpp

```diff
@@ -107,65 +120,45 @@ namespace nmos
-            // It is represented as the base64url encoding of the octet string representation of the coordinate
-            // The "y" (y coordinate) parameter contains the y coordinate for the Elliptic Curve point
-            // It is represented as the base64url encoding of the octet string representation of the coordinate
-            // see https://tools.ietf.org/html/rfc7518#section-6.2.1
-            utility::string_t jwk_to_public_key(const utility::string_t& curve_type, const utility::string_t& base64_x, const utility::string_t& base64_y)
-            {
+#else
                 using ssl::experimental::BIO_ptr;
 
-                // supported Elliptic-Curve types
-                // see https://tools.ietf.org/html/rfc4492#appendix-A
-                const std::map<utility::string_t, int> curve =
-                {
-                    { U("P-256"), NID_X9_62_prime256v1 },
-                    { U("P-384"), NID_secp384r1 },
-                    { U("P-521"), NID_secp521r1 }
-                };
+                auto n = utility::conversions::from_base64url(base64_n);
+                auto e = utility::conversions::from_base64url(base64_e);
+
+                BIGNUM_ptr modulus(BN_bin2bn(n.data(), (int)n.size(), NULL), &BN_free);
+                BIGNUM_ptr exponent(BN_bin2bn(e.data(), (int)e.size(), NULL), &BN_free);
 
-                auto found = curve.find(curve_type);
-                if (curve.end() == found)
+                OSSL_PARAM_BLD_ptr param_bld(OSSL_PARAM_BLD_new(), &OSSL_PARAM_BLD_free);
+                if (OSSL_PARAM_BLD_push_BN(param_bld.get(), OSSL_PKEY_PARAM_RSA_N, modulus.get()))
                 {
-                    throw jwk_exception("convert jwk to pem error: EC type not supported");
+                    modulus.release();
                 }
-                EC_KEY_ptr ec_key(EC_KEY_new_by_curve_name(found->second), &EC_KEY_free);
-                if (!ec_key)
+                if (OSSL_PARAM_BLD_push_BN(param_bld.get(), OSSL_PKEY_PARAM_RSA_E, exponent.get()))
                 {
-                    throw jwk_exception("convert jwk to pem error: failed to create EC key with named curve");
+                    exponent.release();
                 }
 
-                auto x = utility::conversions::from_base64url(base64_x);
-                auto y = utility::conversions::from_base64url(base64_y);
-
-                BIGNUM_ptr x_coordinate(BN_bin2bn(x.data(), (int)x.size(), NULL), &BN_free);
-                BIGNUM_ptr y_coordinate(BN_bin2bn(y.data(), (int)y.size(), NULL), &BN_free);
+                OSSL_PARAM_ptr params(OSSL_PARAM_BLD_to_param(param_bld.get()), &OSSL_PARAM_free);
+                EVP_PKEY_CTX_ptr ctx(EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL), &EVP_PKEY_CTX_free);
 
-                if (EC_KEY_set_public_key_affine_coordinates(ec_key.get(), x_coordinate.get(), y_coordinate.get()))
+                EVP_PKEY* pkey = NULL;
```

Comment:
Would it be possible to use a `EVP_PKEY_ptr` here and avoid the explicit `EVP_PKEY_free` calls below?

Reply:
Yes I had the same thought, but couldn't think how it could be done

Reply:
Done using the boost.ASIO cleanup-struct idiom